src/rust/inetstack/protocols/tcp/established/ctrlblk.rs

```diff
@@ -695,75 +739,13 @@ impl<const N: usize> SharedControlBlock<N> {
-
-            // Advance RCV.NXT over the FIN.
-            self.receiver.receive_next = self.receiver.receive_next + SeqNumber::from(1);
-
-            match self.state {
-                State::Established => self.state = State::CloseWait,
-                State::FinWait1 => {
-                    // RFC 793 has a benign logic flaw.  It says "If our FIN has been ACKed (perhaps in this segment),
-                    // then enter TIME-WAIT, start the time-wait timer, turn off the other timers;".  But if our FIN
-                    // has been ACK'd, we'd be in FIN-WAIT-2 here as a result of processing that ACK (see ACK handling
-                    // above) and will enter TIME-WAIT in the FIN-WAIT-2 case below.  So we can skip that clause and go
-                    // straight to "otherwise enter the CLOSING state".
-                    self.state = State::Closing;
-                },
-                State::FinWait2 => {
-                    // Enter TIME-WAIT.
-                    self.state = State::TimeWait;
-                    // TODO: Start the time-wait timer and turn off the other timers.
-                },
-                State::CloseWait | State::Closing | State::LastAck => (), // Remain in current state.
-                State::TimeWait => {
-                    // TODO: Remain in TIME-WAIT.  Restart the 2 MSL time-wait timeout.
-                },
-                state => panic!("Bad TCP state {:?}", state), // Should never happen.
-            }
-
-            // Push empty buffer.
-            // TODO: set err bit and wake
-            self.receiver.push(DemiBuffer::new(0));
-
-            // Since we consumed the FIN we ACK immediately rather than opportunistically.
-            // TODO: Consider doing this opportunistically.  Note our current tests expect the immediate behavior.
-            self.send_ack();
-
-            return;
+            return Ok(());
         }
 
-        // Check if we need to ACK soon.
-        if should_schedule_ack {
-            // We should ACK this segment, preferably via piggybacking on a response.
-            // TODO: Consider replacing the delayed ACK timer with a simple flag.
-            if self.ack_deadline.get().is_none() {
-                // Start the delayed ACK timer to ensure an ACK gets sent soon even if no piggyback opportunity occurs.
-                let timeout: Duration = self.ack_delay_timeout;
-                self.ack_deadline.set(Some(now + timeout));
-            } else {
-                // We already owe our peer an ACK (the timer was already running), so cancel the timer and ACK now.
-                self.ack_deadline.set(None);
-                self.send_ack();
-            }
-        }
+        // We can only legitimately receive data in ESTABLISHED, FIN-WAIT-1, and FIN-WAIT-2.
+        assert!(self.state == State::Established || self.state == State::FinWait1 || self.state == State::FinWait2);
+        self.receive_data(seg_start, data);
+        Ok(())
```

Comment:
We should still use the match statement. A malfunctioning peer may send us data after the fin, and we don't want to panic, but rather drop the packet.

Reply:
If you look at the code flow, it is no longer possible to enter this function unless we are in the established state, so I removed the assert but certainly I could put a debug assert back.